test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml

```diff
@@ -0,0 +1,157 @@
+  | _ -> ()
+
+let () = Unix.chdir tmp_dir
+
+let () = Path.set_root (Path.External.of_string tmp_dir)
+
+let () = Path.Build.set_build_dir (Path.Build.Kind.of_string "_build")
+
+let now () = Unix.gettimeofday ()
+
+let retry_loop (type a) ~period ~timeout ~(f : unit -> a option) : a option =
+  let t0 = now () in
+  let rec loop () =
+    match f () with
+    | Some res -> Some res
+    | None ->
+      let t1 = now () in
+      if Float.( < ) (t1 -. t0) timeout then (
+        Thread.delay period;
+        loop ()
+      ) else
+        None
+  in
+  loop ()
+
+let print_s sexp = printf "%s\n" (Sexp.to_string_hum sexp)
+
+let sexp_of_path path = sexp_of_string (Path.to_string path)
+
+let get_events ~try_to_get_events ~expected =
+  let collected = ref [] in
+  let done_collecting =
+    match expected with
+    | 0 -> Some `Enough
+    | n ->
+      assert (n > 0);
+      retry_loop ~period:0.01 ~timeout:3.0 ~f:(fun () ->
+          match try_to_get_events () with
+          | None -> None
+          | Some events ->
+            collected := !collected @ events;
+            if List.length !collected >= expected then
+              Some `Enough
+            else
+              None)
+  in
+  match done_collecting with
+  | None -> (!collected, `Not_enough)
+  | Some `Enough ->
+    Thread.delay 0.02;
+    (match try_to_get_events () with
+    | Some events -> collected := !collected @ events
+    | None -> ());
+    if List.length !collected > expected then
+      (!collected, `Too_many)
+    else
+      (!collected, `Ok)
+
+let print_events ~try_to_get_events ~expected =
+  let events, status = get_events ~try_to_get_events ~expected in
+  List.iter events ~f:(fun event -> print_s ([%sexp_of: path] event));
+  match status with
+  | `Ok -> ()
+  | `Not_enough ->
+    printf "Timed out waiting for more events: expected %d, saw %d\n" expected
+      (List.length events)
+  | `Too_many ->
+    printf "Got more events than expected: expected %d, saw %d\n" expected
+      (List.length events)
+
+let%expect_test _ =
+  let mutex = Mutex.create () in
+  let critical_section ~f =
+    Mutex.lock mutex;
+    Exn.protect ~f ~finally:(fun () -> Mutex.unlock mutex)
+  in
+  let events_buffer = ref [] in
+  let watcher =
+    Dune_file_watcher.create ~debounce_interval:None
+      ~thread_safe_send_files_changed:(fun events ->
+        Mutex.lock mutex;
+        events_buffer := !events_buffer @ events;
+        Mutex.unlock mutex)
+      ~root:Path.root
+  in
+  let try_to_get_events () =
+    critical_section ~f:(fun () ->
+        match !events_buffer with
+        | [] -> None
+        | list ->
+          events_buffer := [];
+          Some list)
+  in
+  let print_events n = print_events ~try_to_get_events ~expected:n in
+  Dune_file_watcher.wait_watches_established_blocking watcher;
+  Stdio.Out_channel.write_all "x" ~data:"x";
+  print_events 1;
+  [%expect {|
+x
+|}];
+  (* CR-someday aalekseyev: renaming is not detected *)
```

Comment:
:flushed: never realised we were ignoring all these events. Seems we are not passing enough `--event` arguments to  inotifywait.

Reply:
Yeah, adding a bunch of `--event` arguments could definitely help.